table/server/common/src/main/java/alluxio/table/common/udb/PathTranslator.java

```diff
@@ -66,23 +61,60 @@ public PathTranslator addMapping(String alluxioPath, String ufsPath) {
-    for (BiMap.Entry<String, String> entry : mPathMap.entrySet()) {
-      if (ufsPath.startsWith(entry.getValue())) {
-        // return ufsPath if set the key and value to be same when bypass path.
-        if (entry.getKey().equals(entry.getValue())) {
-          return ufsPath;
-        }
-        String alluxioPath = entry.getKey() + ufsPath.substring(entry.getValue().length());
-        if (alluxioPath.startsWith("/")) {
-          // scheme/authority are missing, so prefix with the scheme and authority
-          alluxioPath =
-              ConfigurationUtils.getSchemeAuthority(ServerConfiguration.global()) + alluxioPath;
+    String suffix = ufsPath.endsWith("/") ? "/" : "";
+    AlluxioURI ufsUri = new AlluxioURI(ufsPath);
+    // first look for an exact match
+    if (mPathMap.inverse().containsKey(ufsUri)) {
+      AlluxioURI match = mPathMap.inverse().get(ufsUri);
+      if (match.equals(ufsUri)) {
+        // bypassed UFS path, return as is
+        return ufsPath;
+      }
+      return checkAndAddSchemeAuthority(mPathMap.inverse().get(ufsUri)) + suffix;
+    }
+    // otherwise match by longest prefix
+    BiMap.Entry<AlluxioURI, AlluxioURI> longestPrefix = null;
+    int longestPrefixDepth = -1;
+    for (BiMap.Entry<AlluxioURI, AlluxioURI> entry : mPathMap.entrySet()) {
+      try {
+        AlluxioURI valueUri = entry.getValue();
+        if (valueUri.isAncestorOf(ufsUri) && valueUri.getDepth() > longestPrefixDepth) {
+          longestPrefix = entry;
+          longestPrefixDepth = valueUri.getDepth();
         }
-        return alluxioPath;
+      } catch (InvalidPathException e) {
+        throw new IOException(e);
       }
     }
-    // TODO(yuzhu): instead of throwing an exception, mount the path?
-    throw new IOException(String
-        .format("Failed to translate ufs path (%s). Mapping missing from translator", ufsPath));
+    if (longestPrefix == null) {
+      // TODO(yuzhu): instead of throwing an exception, mount the path?
+      throw new IOException(String
+          .format("Failed to translate ufs path (%s). Mapping missing from translator", ufsPath));
+    }
+    if (longestPrefix.getKey().equals(longestPrefix.getValue())) {
+      // return ufsPath if set the key and value to be same when bypass path.
+      return ufsPath;
+    }
+    try {
+      String difference = PathUtils.subtractPaths(ufsUri.getPath(),
+          longestPrefix.getValue().getPath());
+      AlluxioURI mappedUri = longestPrefix.getKey().join(difference);
+      return checkAndAddSchemeAuthority(mappedUri) + suffix;
+    } catch (InvalidPathException e) {
+      throw new IOException(e);
+    }
+  }
+
+  private AlluxioURI checkAndAddSchemeAuthority(AlluxioURI input) {
+    if (!input.hasScheme()) {
+      AlluxioURI baseUri = new AlluxioURI(getSchemeAuthority() + "/");
+      return new AlluxioURI(baseUri, input.getPath(), false);
+    }
+    return input;
+  }
+
+  @VisibleForTesting
+  String getSchemeAuthority() {
```

Comment:
since there is no member variable used, and this is changed over time, can we change this method to be a `static` variable instead, to save computation on every call?